metrics/scaling/k8s_scale.sh

```diff
@@ -0,0 +1,267 @@
+	cpu_idle=${cpu_idle:-0}
+	mem_free=${mem_free:-0}
+
+	local json="$(cat << EOF
+	{
+		"n_pods": {
+			"Result": ${n_pods},
+			"Units" : "int"
+		},
+		"cpu_idle": {
+			"Result": ${cpu_idle},
+			"Units" : "%"
+		},
+		"launch_time": {
+			"Result": $launch_time,
+			"Units" : "s"
+		},
+		"mem_free": {
+			"Result": ${mem_free},
+			"Units" : "kb"
+		}
+	}
+EOF
+)"
+	metrics_json_add_array_element "$json"
+}
+
+init() {
+	info "Initialising"
+	info "Checking k8s accessible"
+	local worked=$( kubectl get nodes > /dev/null 2>&1 && echo $? || echo $? )
+	if [ "$worked" != 0 ]; then
+		die "kubectl failed to get nodes"
+	fi
+
+	info $(get_num_nodes) "k8s nodes found"
+	# We could check we have just the one node here - right now this is a single node
+	# test!! - because, our stats gathering is rudimentry, as k8s does not provide
+	# a nice way to do it (unless you want to parse 'descibe nodes')
+	# Have a read of https://github.com/kubernetes/kubernetes/issues/25353
+
+	# FIXME - check the node(s) can run enough pods - check 'max-pods' in the
+	# kubelet config - from 'kubectl describe node -o json' ?
+
+	# Launch our stats gathering pod
+	kubectl apply -f ${stats_pod}.yaml
+	kubectl wait --for=condition=Ready pod "${stats_pod}"
+
+	# FIXME - we should probably 'warm up' the cluster with the container image(s) we will
+	# use for testing, otherwise the download time will likely be included in the first pod
+	# boot time.
+
+	# And now we can set up our results storage then...
+	metrics_json_init "k8s"
+	save_config
+}
+
+save_config(){
+	metrics_json_start_array
+
+	local json="$(cat << EOF
+	{
+		"testname": "${TEST_NAME}",
+		"NUM_PODS": "${NUM_PODS}",
+		"STEP": "${STEP}",
+		"wait_time": "${wait_time}",
+		"delete_wait_time": "${delete_wait_time}",
+		"settle_time": "${settle_time}"
+	}
+EOF
+)"
+	metrics_json_add_array_element "$json"
+	metrics_json_end_array "Config"
+}
+
+run() {
+	info "Running test"
+
+	trap cleanup EXIT QUIT KILL
+
+	metrics_json_start_array
+	for reqs in $(seq ${STEP} ${STEP} ${NUM_PODS}); do
+		info "Testing replicas ${reqs} of ${NUM_PODS}"
+		# Generate the next yaml file
+
+	if [ "$use_kata_runtime" != "no" ]; then
+		sed -e "s|@REPLICAS@|${reqs}|g" \
+		    -e "s|@RUNTIMECLASS@|${RUNTIME}|g" \
+		    -e "s|@DEPLOYMENT@|${deployment}|g" \
+		    -e "s|@LABEL@|${LABEL}|g" \
+		    -e "s|@LABELVALUE@|${LABELVALUE}|g" \
+		    < ${input_yaml} > ${generated_yaml}
+    else
+        sed -e "s|@REPLICAS@|${reqs}|g" \
+		    -e "/@RUNTIMECLASS@/d" \
+		    -e "s|@DEPLOYMENT@|${deployment}|g" \
+		    -e "s|@LABEL@|${LABEL}|g" \
+		    -e "s|@LABELVALUE@|${LABELVALUE}|g" \
+		    < ${input_yaml} > ${generated_yaml}
+    fi
+
+		info "Applying changes"
+		# FIXME - time taken in 'seconds' - we can do better than that, if we
+		# use the date nanosecond stamp, and steal the magic from the density test script.
+        # switch to 'date +%N'
+		local start_time=$(date +%s)
```

Comment:
the json.sh lib file has ms and ns time functions we can now use here I reckon..

Reply:
changed to milliseconds